sdk/nodejs/policy/policy.ts

```diff
@@ -82,10 +69,142 @@ export interface Policy {
      * proper permissions.
      */
     enforcementLevel: EnforcementLevel;
+}
+
+/**
+ * An array of Policies.
+ */
+export type Policies = (ResourceValidationPolicy | StackValidationPolicy)[];
+
+/**
+ * ResourceValidationPolicy is a policy that validates a resource definition.
+ */
+export interface ResourceValidationPolicy extends Policy {
+    /**
+     * A callback function that validates if a resource definition violates a policy (e.g. "S3 buckets
+     * can't be public"). A single callback function can be specified, or multiple functions, which are
+     * called in order.
+     */
+    validateResource: ResourceValidation | ResourceValidation[];
+}
+
+/**
+ * ResourceValidation is the callback signature for a `ResourceValidationPolicy`. A resource validation
+ * is passed `args` with more information about the resource and a `reportViolation` callback that can be
+ * used to report a policy violation. `reportViolation` can be called multiple times to report multiple
+ * violations against the same resource. `reportViolation` must be passed a message about the violation.
+ * The `reportViolation` signature accepts an optional `urn` argument, which is ignored when validating
+ * resources (the `urn` of the resource being validated is always used).
+ */
+export type ResourceValidation = (args: ResourceValidationArgs, reportViolation: ReportViolation) => Promise<void> | void;
+
+/**
+ * ResourceValidationArgs is the argument bag passed to a resource validation.
+ */
+export interface ResourceValidationArgs {
+    /**
+     * The type of the Resource.
+     */
+    type: string;
+
+    /**
+     * The properties of the Resource.
+     */
+    props: Record<string, any>;
+
+    // TODO: Add support for the following:
+    //
+    // urn: string;
+    // name: string;
+    // opts: PolicyResourceOptions;
+}
+
+/**
+ * A helper function that returns a strongly-typed resource validation function.
+ * @param filter A type guard used to determine if the args are an instance of the resource.
```

Comment:
Would `typeFilter` or `typeGuard` make more sense here?

Reply:
Let's go with `typeFilter` which is also the naming Alex used [here](https://github.com/pulumi/pulumi/blob/23a3acc46f0df6deec4443d0f107400311c2f960/sdk/nodejs/runtime/resource.ts#L563).